### PR TITLE
feat: Add Media Recorder

### PR DESCRIPTION
**Pull request by istarkov**
## Description

Media recorder hook and useLongPressToggle hook.

https://6382151c8b47d4399fb9fc69-gumktzhrtx.chromatic.com/?path=/story/library-media-recorder--media-recorder

## Steps for reproduction

See storybook for usage.

## Code Review

- [ ] hi @kof, I need you to do
  - conceptual review (architecture, feature-correctness)
  - detailed review (read every line)
  - test it on preview

## Before requesting a review

- [ ] made a self-review
- [ ] added inline comments where things may be not obvious (the "why", not "what")

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio/blob/main/apps/builder/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env.example` and the `builder/env-check.js` if mandatory

### CODE REVIEW

File: apps/builder/app/builder/features/ai/hooks/effect-event.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+// eslint-disable-next-line @typescript-eslint/ban-types
+export const useEffectEvent = <T extends Function>(callback?: T) => {
+  const ref = useRef(callback);
+
+  useSafeInsertionEffect(() => {
+    ref.current = callback;
+  });
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (TrySound) [inline apps/builder/app/builder/features/ai/hooks/effect-event.ts:18]: Why does it fail? Refs can be excluded from deps. Linter knows this.

File: apps/builder/app/builder/features/ai/hooks/long-press-toggle.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+ * Toggle hook with long-press handling.
+ * - `onStart`: First click.
+ * - `onEnd`: Second click or long-press release.
+ * - `onCancel`: Pointer up outside target during long press.
+ */
+export const useLongPressToggle = (props: UseClickAndHoldProps) => {
+  const currentTarget = useRef<Element>();
+  const pointerDownTimeRef = useRef(0);
+  const stateRef = useRef<"idle" | "active">("idle");
+  const keyMapRef = useRef<string[]>([]);
```
Comment (TrySound) [inline apps/builder/app/builder/features/ai/hooks/long-press-toggle.ts:21]: why not use set?

File: apps/builder/app/builder/features/ai/hooks/effect-event.ts (hunk 1 above)
Comment (istarkov) [reply apps/builder/app/builder/features/ai/hooks/effect-event.ts:18]: image

File: apps/builder/app/builder/features/ai/hooks/media-recorder.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
+/* eslint-disable */
```
Comment (TrySound) [inline apps/builder/app/builder/features/ai/hooks/media-recorder.ts:1]: Why suppress everything?

Comment (istarkov) [reply apps/builder/app/builder/features/ai/hooks/media-recorder.ts:1]: Forget to delete, during development eslint have too many rules I don't want to think about

File: apps/builder/app/builder/features/ai/hooks/long-press-toggle.ts (hunk 2 above)
Comment (istarkov) [reply apps/builder/app/builder/features/ai/hooks/long-press-toggle.ts:21]: 2 elements inside, array is faster ;-)

Comment (istarkov) [reply apps/builder/app/builder/features/ai/hooks/long-press-toggle.ts]: Switched on Set

File: apps/builder/app/builder/features/ai/hooks/effect-event.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+const useSafeInsertionEffect =
+  typeof window !== "undefined"
+    ? useInsertionEffect
+    : (fn: () => void) => {
+        fn();
+      };
+
+// eslint-disable-next-line @typescript-eslint/ban-types
+export const useEffectEvent = <T extends Function>(callback?: T) => {
```
Comment (kof) [inline apps/builder/app/builder/features/ai/hooks/effect-event.ts:11]: we do have hook-utils folder

File: apps/builder/app/builder/features/ai/hooks/media-recorder.ts [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    // New recording started, do cleanup and return
+    if (id !== idRef.current) {
+      stream.getAudioTracks().forEach((track) => track.stop());
+      return;
+    }
+
+    setError(undefined);
+    setState("recording");
+
+    const extension = MediaRecorder.isTypeSupported("audio/webm; codecs=opus")
```
Comment (kof) [inline apps/builder/app/builder/features/ai/hooks/media-recorder.ts:70]: nit:
```suggestion
    const format = MediaRecorder.isTypeSupported("audio/webm; codecs=opus")
```

**Review by kof — APPROVED** [review]
nice!

File: apps/builder/app/builder/features/ai/hooks/media-recorder.ts (hunk 5 above)
Comment (istarkov) [reply apps/builder/app/builder/features/ai/hooks/media-recorder.ts:70]: subtype is the right name, https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types#structure_of_a_mime_type
renamed

**Review by TrySound — APPROVED** [review]

**Pull request merged**